File: relicframe/riven_trade_chat.py
"""Local, explicit trade-chat offer importer.

Warframe does not expose a public trade-chat feed.  This module only parses
text the user deliberately supplies (for example, text copied from Windows
Snipping Tool).  WTS/WTB messages are observations, never recorded as sales.
"""

from __future__ import annotations

import hashlib
import json
import re
import statistics
import threading
from dataclasses import asdict, dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Iterable


ACTION_RE = re.compile(r"\b(WTS|WTB|WTT)\b", re.IGNORECASE)
PRICE_RE = re.compile(r"(?<!\d)(\d{1,6})\s*(?:p|pl|plat|platinum)\b", re.IGNORECASE)
TIME_PREFIX_RE = re.compile(r"^\s*(?:\[\s*\d{1,2}:\d{2}(?::\d{2})?\s*\]\s*)+")
BRACKET_RE = re.compile(r"\[([^\[\]]{2,160})\]")


def _key(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", " ", value.casefold()).strip()


@dataclass(frozen=True)
class TradeChatOffer:
    offer_id: str
    observed_at: str
    action: str
    seller: str
    weapon: str
    price: int | None
    raw_text: str
    source: str = "manual"


@dataclass(frozen=True)
class TradeChatImportResult:
    added: tuple[TradeChatOffer, ...]
    duplicate_count: int
    unparsed_line_count: int


TRADE_CHAT_BUDGET_BYTES = 16 * 1024 * 1024


class RivenTradeChatLog:
    def __init__(self, path: str | Path, weapon_names: Iterable[str]):
        self.path = Path(path)
        self.weapon_names = tuple(sorted(
            {name.strip() for name in weapon_names if name and name.strip()},
            key=lambda value: len(_key(value)),
            reverse=True,
        ))
        self._lock = threading.Lock()

    def _weapon_in(self, value: str) -> str | None:
        normalized = f" {_key(value)} "
        for weapon in self.weapon_names:
            marker = f" {_key(weapon)} "
            if marker in normalized or normalized.strip().startswith(marker.strip() + " "):
                return weapon
        return None

    @staticmethod
    def _seller(line: str, action_start: int) -> str:
        prefix = TIME_PREFIX_RE.sub("", line[:action_start]).strip(" :-|>")
        if ":" in prefix:
            prefix = prefix.rsplit(":", 1)[-1].strip()
        prefix = re.sub(r"^[^A-Za-z0-9_.-]+", "", prefix)
        return prefix[-32:] if prefix else "Unknown"

    def parse(self, text: str, *, observed_at: datetime | None = None, source: str = "manual") -> tuple[list[TradeChatOffer], int]:
        stamp = (observed_at or datetime.now(UTC)).astimezone(UTC)
        offers: list[TradeChatOffer] = []
        unparsed = 0
        for raw in text.splitlines():
            line = " ".join(raw.strip().split())
            if not line:
                continue
            action_match = ACTION_RE.search(line)
            if not action_match:
                unparsed += 1
                continue
            action = action_match.group(1).upper()
            seller = self._seller(line, action_match.start())
            brackets = list(BRACKET_RE.finditer(line))
            found: list[tuple[str, int | None]] = []
            for index, bracket in enumerate(brackets):
                weapon = self._weapon_in(bracket.group(1))
                if not weapon:
                    continue
                end = brackets[index + 1].start() if index + 1 < len(brackets) else len(line)
                nearby = PRICE_RE.search(line[bracket.end():end])
                found.append((weapon, int(nearby.group(1)) if nearby else None))
            if not found:
                weapon = self._weapon_in(line[action_match.end():])
                if weapon:
                    price_match = PRICE_RE.search(line[action_match.end():])
                    found.append((weapon, int(price_match.group(1)) if price_match else None))
            if not found:
                unparsed += 1
                continue
            global_price = PRICE_RE.search(line[action_match.end():])
            for weapon, price in found:
                resolved_price = price if price is not None else (int(global_price.group(1)) if global_price else None)
                identity = "\0".join((stamp.date().isoformat(), action, _key(seller), _key(weapon), str(resolved_price), _key(line)))
                offer_id = hashlib.sha256(identity.encode("utf-8")).hexdigest()[:20]
                offers.append(TradeChatOffer(
                    offer_id=offer_id,
                    observed_at=stamp.isoformat(),
                    action=action,
                    seller=seller,
                    weapon=weapon,
                    price=resolved_price,
                    raw_text=line,
                    source=source,
                ))
        return offers, unparsed

    def load(self) -> list[TradeChatOffer]:
        return list(self._iter_offers())

    def _iter_offers(self):
        if not self.path.exists():
            return
        try:
            with self.path.open(encoding="utf-8") as stream:
                for line in stream:
                    try:
                        yield TradeChatOffer(**json.loads(line))
                    except (json.JSONDecodeError, TypeError, ValueError):
                        continue
        except OSError:
            return

    def import_text(self, text: str, *, observed_at: datetime | None = None, source: str = "manual") -> TradeChatImportResult:
        parsed, unparsed = self.parse(text, observed_at=observed_at, source=source)
        with self._lock:
            existing_ids = {offer.offer_id for offer in self._iter_offers()}
            added = [offer for offer in parsed if offer.offer_id not in existing_ids]
            duplicates = len(parsed) - len(added)
            if added:
                size = self.path.stat().st_size if self.path.exists() else 0
                payload_size = sum(len((json.dumps(asdict(offer), ensure_ascii=False) + "\n").encode("utf-8")) for offer in added)
                if size + payload_size > TRADE_CHAT_BUDGET_BYTES:
                    raise RuntimeError("Trade-chat log reached its 16 MiB budget. Export/archive it before importing more; existing data is unchanged.")
                self.path.parent.mkdir(parents=True, exist_ok=True)
                with self.path.open("a", encoding="utf-8", newline="\n") as target:
                    for offer in added:
                        target.write(json.dumps(asdict(offer), ensure_ascii=False) + "\n")
        return TradeChatImportResult(tuple(added), duplicates, unparsed)

    def recent(self, *, weapon: str | None = None, days: int = 30) -> list[TradeChatOffer]:
        cutoff = datetime.now(UTC) - timedelta(days=max(1, days))
        marker = _key(weapon or "")
        result = []
        for offer in self._iter_offers():
            try:
                stamp = datetime.fromisoformat(offer.observed_at)
            except ValueError:
                continue
            if stamp.tzinfo is None:
                stamp = stamp.replace(tzinfo=UTC)
            if stamp < cutoff or (marker and _key(offer.weapon) != marker):
                continue
            result.append(offer)
        return sorted(result, key=lambda offer: offer.observed_at, reverse=True)


def price_summary(offers: Iterable[TradeChatOffer]) -> dict[str, float | int | None]:
    rows = list(offers)
    wts = [offer.price for offer in rows if offer.action == "WTS" and offer.price is not None]
    wtb = [offer.price for offer in rows if offer.action == "WTB" and offer.price is not None]
    return {
        "observations": len(rows),
        "wts_count": len(wts),
        "wts_min": min(wts) if wts else None,
        "wts_median": statistics.median(wts) if wts else None,
        "wtb_count": len(wtb),
        "wtb_max": max(wtb) if wtb else None,
        "wtb_median": statistics.median(wtb) if wtb else None,
        "possible_spread": (max(wtb) - min(wts)) if wts and wtb else None,
    }

File: csharp/RelicFrame.Core/RivenTradeChat.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace RelicFrame.Core;

public sealed record TradeChatOffer(string OfferId, string ObservedAt, string Action, string Seller, string Weapon,
    int? Price, string RawText, string Source = "manual");
public sealed record TradeChatImport(TradeChatOffer[] Added, int DuplicateCount, int UnparsedLineCount);
public sealed record TradeChatSummary(int Observations, int WtsCount, double? WtsMin, double? WtsMedian,
    int WtbCount, double? WtbMax, double? WtbMedian, double? PossibleSpread);

public sealed class RivenTradeChat
{
    private static readonly Regex ActionPattern = new("\\b(WTS|WTB|WTT)\\b", RegexOptions.IgnoreCase | RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
    private static readonly Regex PricePattern = new("(?<!\\d)(\\d{1,6})\\s*(?:p|pl|plat|platinum)\\b", RegexOptions.IgnoreCase | RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
    private static readonly Regex TimePrefix = new("^\\s*(?:\\[\\s*\\d{1,2}:\\d{2}(?::\\d{2})?\\s*\\]\\s*)+", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
    private static readonly Regex Brackets = new("\\[([^\\[\\]]{2,160})\\]", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
    private readonly string path;
    private readonly Func<IReadOnlyList<string>> weaponSource;
    private readonly object weaponGate = new();
    private IReadOnlyList<string>? weaponSnapshot;
    private string[] weapons = [];
    private readonly object gate = new();
    private readonly List<TradeChatOffer> cached = [];
    private readonly HashSet<string> cachedIds = new(StringComparer.Ordinal);
    private bool loaded;
    public RivenTradeChat(string path, IEnumerable<string> weapons)
    {
        this.path = path; var fixedWeapons = weapons.ToArray(); weaponSource = () => fixedWeapons;
        weaponSnapshot = fixedWeapons; this.weapons = NormalizeWeapons(fixedWeapons);
    }
    public RivenTradeChat(string path, Func<IReadOnlyList<string>> weaponSource)
    { this.path = path; this.weaponSource = weaponSource ?? throw new ArgumentNullException(nameof(weaponSource)); }
    private static string[] NormalizeWeapons(IEnumerable<string> source) => source.Where(w => !string.IsNullOrWhiteSpace(w)).Select(w => w.Trim())
        .Distinct(StringComparer.OrdinalIgnoreCase).OrderByDescending(w => Key(w).Length).ToArray();
    private string[] CurrentWeapons()
    {
        var snapshot = weaponSource();
        if (ReferenceEquals(snapshot, Volatile.Read(ref weaponSnapshot))) return Volatile.Read(ref weapons);
        lock (weaponGate)
        {
            if (!ReferenceEquals(snapshot, weaponSnapshot))
            {
                Volatile.Write(ref weapons, NormalizeWeapons(snapshot));
                Volatile.Write(ref weaponSnapshot, snapshot);
            }
            return weapons;
        }
    }
    private static string Key(string value) => Regex.Replace(value.ToLowerInvariant(), "[^a-z0-9]+", " ").Trim();
    private string? WeaponIn(string value)
    {
        var normalized = " " + Key(value) + " ";
        return CurrentWeapons().FirstOrDefault(w => normalized.Contains(" " + Key(w) + " ", StringComparison.Ordinal)
            || normalized.Trim().StartsWith(Key(w) + " ", StringComparison.Ordinal));
    }
    private static string Seller(string line, int actionStart)
    {
        var prefix = TimePrefix.Replace(line[..actionStart], "").Trim(' ', ':', '-', '|', '>');
        if (prefix.Contains(':')) prefix = prefix[(prefix.LastIndexOf(':') + 1)..].Trim();
        prefix = Regex.Replace(prefix, "^[^A-Za-z0-9_.-]+", ""); return prefix.Length == 0 ? "Unknown" : prefix[^Math.Min(32, prefix.Length)..];
    }
    public (TradeChatOffer[] Offers, int Unparsed) Parse(string text, DateTimeOffset? observed = null, string source = "manual")
    {
        if (text.Length > 1_000_000) throw new InvalidDataException("One manual trade-chat import is limited to 1 MiB.");
        var stamp = (observed ?? DateTimeOffset.UtcNow).ToUniversalTime();
        var timestamp = stamp.Ticks % TimeSpan.TicksPerSecond == 0 ? stamp.ToString("yyyy-MM-dd'T'HH:mm:sszzz") : stamp.ToString("yyyy-MM-dd'T'HH:mm:ss.ffffffzzz");
        var offers = new List<TradeChatOffer>(); var unparsed = 0;
        foreach (var raw in text.Replace("\r", "").Split('\n'))
        {
            var line = string.Join(' ', raw.Trim().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)); if (line.Length == 0) continue;
            var action = ActionPattern.Match(line); if (!action.Success) { unparsed++; continue; }
            var kind = action.Groups[1].Value.ToUpperInvariant(); var seller = Seller(line, action.Index); var brackets = Brackets.Matches(line).Cast<Match>().ToArray();
            var found = new List<(string Weapon, int? Price)>();
            for (var i = 0; i < brackets.Length; i++)
            {
                var weapon = WeaponIn(brackets[i].Groups[1].Value); if (weapon is null) continue;
                var end = i + 1 < brackets.Length ? brackets[i + 1].Index : line.Length; var near = PricePattern.Match(line, brackets[i].Index + brackets[i].Length, end - brackets[i].Index - brackets[i].Length);
                found.Add((weapon, near.Success ? int.Parse(near.Groups[1].Value) : null));
            }
            if (found.Count == 0 && WeaponIn(line[(action.Index + action.Length)..]) is { } fallbackWeapon)
            { var match = PricePattern.Match(line, action.Index + action.Length); found.Add((fallbackWeapon, match.Success ? int.Parse(match.Groups[1].Value) : null)); }
            if (found.Count == 0) { unparsed++; continue; }
            var global = PricePattern.Match(line, action.Index + action.Length);
            foreach (var item in found)
            {
                var price = item.Price ?? (global.Success ? int.Parse(global.Groups[1].Value) : null);
                var identity = string.Join('\0', stamp.ToString("yyyy-MM-dd"), kind, Key(seller), Key(item.Weapon), price?.ToString() ?? "None", Key(line));
                var id = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(identity))).ToLowerInvariant()[..20];
                offers.Add(new(id, timestamp, kind, seller, item.Weapon, price, line, source));
            }
        }
        return (offers.ToArray(), unparsed);
    }
    private void EnsureLoaded()
    {
        if (loaded) return;
        if (File.Exists(path)) foreach (var line in File.ReadLines(path))
        {
            try
            {
                if (JsonSerializer.Deserialize<TradeChatOffer>(line, Json.Options) is not { } row || !cachedIds.Add(row.OfferId)) continue;
                cached.Add(row);
            }
            catch (JsonException) { }
        }
        loaded = true;
    }
    public TradeChatOffer[] Load()
    {
        lock (gate) { EnsureLoaded(); return cached.ToArray(); }
    }
    public TradeChatImport Import(string text, DateTimeOffset? observed = null, string source = "manual")
    {
        var parsed = Parse(text, observed, source);
        lock (gate)
        {
            EnsureLoaded(); var added = parsed.Offers.Where(o => cachedIds.Add(o.OfferId)).ToArray();
            var payload = added.Select(o => JsonSerializer.Serialize(o, Json.Options) + "\n").ToArray(); var bytes = payload.Sum(Encoding.UTF8.GetByteCount);
            if ((File.Exists(path) ? new FileInfo(path).Length : 0) + bytes > 16 * 1024 * 1024)
            {
                foreach (var row in added) cachedIds.Remove(row.OfferId);
                throw new InvalidDataException("Trade-chat log reached 16 MiB; archive it before importing more. Existing data is unchanged.");
            }
            if (added.Length > 0) try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(path))!);
                using var output = new StreamWriter(path, true, new UTF8Encoding(false)); foreach (var line in payload) output.Write(line);
            }
            catch
            {
                foreach (var row in added) cachedIds.Remove(row.OfferId);
                throw;
            }
            cached.AddRange(added);
            return new(added, parsed.Offers.Length - added.Length, parsed.Unparsed);
        }
    }
    public TradeChatOffer[] Recent(string? weapon = null, int days = 30)
    {
        var cutoff = DateTimeOffset.UtcNow.AddDays(-Math.Max(1, days)); var key = Key(weapon ?? "");
        return Load().Where(o => DateTimeOffset.TryParse(o.ObservedAt, out var stamp) && stamp >= cutoff && (key.Length == 0 || Key(o.Weapon) == key)).OrderByDescending(o => o.ObservedAt, StringComparer.Ordinal).ToArray();
    }
    public static TradeChatSummary Summary(IEnumerable<TradeChatOffer> rows)
    {
        var all = rows.ToArray(); var wts = all.Where(o => o.Action == "WTS" && o.Price.HasValue).Select(o => (double)o.Price!.Value).Order().ToArray();
        var wtb = all.Where(o => o.Action == "WTB" && o.Price.HasValue).Select(o => (double)o.Price!.Value).Order().ToArray();
        static double? Median(double[] a) => a.Length == 0 ? null : a.Length % 2 > 0 ? a[a.Length / 2] : (a[a.Length / 2 - 1] + a[a.Length / 2]) / 2;
        return new(all.Length, wts.Length, wts.FirstOrDefault(double.NaN) is var min && double.IsFinite(min) ? min : null, Median(wts),
            wtb.Length, wtb.LastOrDefault(double.NaN) is var max && double.IsFinite(max) ? max : null, Median(wtb), wts.Length > 0 && wtb.Length > 0 ? wtb.Max() - wts.Min() : null);
    }
}
